xiilib/_gunicorn/charm_state.py

```diff
@@ -0,0 +1,217 @@
+        }
+        app_config = {k: v for k, v in app_config.items() if k not in wsgi_config.dict().keys()}
+        return cls(
+            framework=framework,
+            wsgi_config=wsgi_config.dict(exclude_unset=True, exclude_none=True),
+            app_config=typing.cast(dict[str, str | int | bool], app_config),
+            database_requirers=database_requirers,
+            webserver_config=WebserverConfig.from_charm(charm),
+            secret_key=(
+                secret_storage.get_secret_key() if secret_storage.is_initialized else None
+            ),
+            is_secret_storage_ready=secret_storage.is_initialized,
+        )
+
+    @property
+    def proxy(self) -> "ProxyConfig":
+        """Get charm proxy information from juju charm environment.
+
+        Returns:
+            charm proxy information in the form of `ProxyConfig`.
+        """
+        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY")
+        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY")
+        no_proxy = os.environ.get("JUJU_CHARM_NO_PROXY")
+        return ProxyConfig(
+            http_proxy=parse_obj_as(AnyHttpUrl, http_proxy) if http_proxy else None,
+            https_proxy=parse_obj_as(AnyHttpUrl, https_proxy) if https_proxy else None,
+            no_proxy=no_proxy,
+        )
+
+    @property
+    def wsgi_config(self) -> dict[str, str | int | bool]:
+        """Get the value of the WSGI application specific configuration.
+
+        Returns:
+            The value of the WSGI application specific configuration.
+        """
+        return self._wsgi_config
+
+    @property
+    def app_config(self) -> dict[str, str | int | bool]:
+        """Get the value of user-defined application configurations.
+
+        Returns:
+            The value of user-defined application configurations.
+        """
+        return self._app_config
+
+    @property
+    def port(self) -> int:
+        """Gets the port number to use for the Gunicorn server.
+
+        Returns:
+            The port number to use for the Gunicorn server.
+        """
+        return 8000
+
+    @property
+    def statsd_host(self) -> str:
+        """Returns the statsd server host for Gunicorn metrics.
+
+        Returns:
+            The statsd server host for Gunicorn metrics.
+        """
+        return "localhost:9125"
+
+    @property
+    def secret_key(self) -> str:
+        """Return the WSGI application secret key stored in the SecretStorage.
+
+        It's an error to read the secret key before SecretStorage is initialized.
+
+        Returns:
+            The WSGI application secret key stored in the SecretStorage.
+
+        Raises:
+            RuntimeError: raised when accessing WSGI application secret key before
+                          secret storage is ready.
+        """
+        if self._secret_key is None:
+            raise RuntimeError("access secret key before secret storage is ready")
```

Comment:
Shouldn't we raise a custom error ?

Reply:
This is a defensive programming, it will not raise if the code is written correctly.

Reply:
In such case I think it's fine then